base/set.jl

```diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(c)
+Remove duplicate items and return the modified collection, as determined by
+[`isequal`](@ref).
+
+```jldoctest
+julia> unique!([1, 1, 1])
+1-element Array{Int64,1}:
+ 1
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
+
+julia> A
+4-element Array{Int64,1}:
+ 7
+ 3
+ 2
+ 5
+```
+"""
+function unique!(c)
+    if issorted(c) || issorted(reverse(c))
+        idxs = eachindex(c)
+        # When we loop through c, we only need to remember one state and one
+        # item in the collection.
+        currentstate = start(idxs)
+        currentitem = c[currentstate]
```

Comment:
This will fail for `unique!([])`.  Manually using the iteration protocol is a little tedious, but you always need to call `done(idxs, currentstate)` before getting the next element with `idx, currentstate = next(idxs, currentstate)`. And then you can't really use `currentstate` for anything beyond those two functions (`done` and `next`).  It's an internal implementation detail… and you can't count on it being useful as an index itself.